### PR TITLE
Content for .NET Aspire - Release 8.1

### LINKED ISSUE #1039: [Components] New Milvus component article
### Describe the issue or suggestion

**Help us make content visible**

**Describe the new article**

We added a new `Aspire.Milvus.Client` component and `Aspire.Hosting.Milvus` hosting library in `main`. See:

* https://github.com/dotnet/aspire/issues/796
* https://github.com/dotnet/aspire/pull/4179
* https://github.com/dotnet/aspire/tree/main/src/Components/Aspire.Milvus.Client

cc @timheuer @IEvangelist

### LINKED ISSUE #1048: [New article]: Article for incoming Python support in .NET Aspire 8.1
### Proposed topic or title

Build .NET Aspire apps with Python

### Location in table of contents.

Getting started (similar to Node.js, but suggestions welcome)

### Reason for the article

In .NET Aspire 8.1 we are adding support to start Python executables as a first-class API. This is the PR:

https://github.com/dotnet/aspire/pull/4142

### Article abstract

Basically, the same as the Node.js application but for Python instead.

### Relevant searches

python

---
[Associated WorkItem - 288125](https://dev.azure.com/msft-skilling/Content/_workitems/edit/288125)

### LINKED ISSUE #1049: [New article]: Redis-like article but for Valkey + Aspire.StackExchange.Redis
### Proposed topic or title

Using Valkey with .NET Aspire

### Location in table of contents.

Next to the Redis articles

### Reason for the article

We are introducing support for Valkey in this PR:

https://github.com/dotnet/aspire/pull/4324

It is presently a drop in alternative for Redis. We probably need an article that sites right alongside the Redis articles so that folks who might be looking for Redis content to understand how Valkey works can find it.

### Article abstract

The article should cover:

1. AddValkey
2. WithDataVolume/WithBindMount
3. WithPersistence

It can probably cover off the client side pieces by just referring back to the Redis article although the Redis article might need to more clearly call out which is service code and which is app host code.

Configuring Valkey using AddValkey and using the data volume and persistence extension methods.

### Relevant searches

valkey, redis

---
[Associated WorkItem - 288126](https://dev.azure.com/msft-skilling/Content/_workitems/edit/288126)

### LINKED ISSUE #1059: [New article]: Article for incoming Elasticsearch Hosting in .NET Aspire 8.1
### Proposed topic or title

Add Elasticsearch in .NET aspire apps

### Location in table of contents.

Database

### Reason for the article

in .NET Aspire 8.1 we are adding support to Elasticsearch hosting as a first-class API. See:

> * [Add Elasticsearch Component aspire#4418](https://github.com/dotnet/aspire/issues/4418)
> * [Add Elasticsearch Hosting aspire#4430](https://github.com/dotnet/aspire/pull/4430)

### Article abstract

The article should cover:

AddElasticsearch
WithDataVolume/WithDataBindMount
for elastic hosting

also, its should cover client component when its ready

### Relevant searches

elasticsearch

---
[Associated WorkItem - 288127](https://dev.azure.com/msft-skilling/Content/_workitems/edit/288127)

### LINKED ISSUE #1063: [New article]: Conceptual docs for WithDockerfile
### Proposed topic or title

Adding Dockerfiles to the .NET Aspire application model

### Location in table of contents.

_No response_

### Reason for the article

We are adding a new extension method called `WithDockerfile` which allows .NET Aspire to dynamically build a container image from a Dockerfile and run it.

### Article abstract

Needs to cover these things:

1. Basics of getting it working.
2. Using build arguments
3. Using build secrets
4. Using to augment existing custom container resources (e.g. `AddSqlServer(...)`).
5. Troubleshooting (example of images not being rebuilt)

### Relevant searches

_No response_

---
[Associated WorkItem - 288128](https://dev.azure.com/msft-skilling/Content/_workitems/edit/288128)

### LINKED ISSUE #1378: New Keycloak article
We added 2 new libraries to Aspire:

* Aspire.Hosting.Keycloak
* Aspire.Keycloak.Authentication

See https://github.com/dotnet/aspire/pull/4289 for more details.

We should add an article on Keycloak and how to use it in .NET Aspire.

### LINKED ISSUE #1394: [New article]: We need an Azure Web Pub Sub article for Aspire.
### Proposed topic or title

.NET Aspire support for Azure Web PubSub Service

### Location in table of contents.

messaging

### Reason for the article

Was added in 8.1 but we don't have conceptual documents covering it.

### Article abstract

Overall the same as SignalR

### Relevant searches

_No response_

### LINKED ISSUE #1396: Upgrade all snippet bits to release 8.1
### Describe the issue or suggestion

Upgrade all snippet bits to release 8.1

### PR DESCRIPTION
**Pull request by IEvangelist**
## Summary

Content updates for .NET Aspire - Release 8.1

- Fixes #1063
- Fixes #1059
- Fixes #1049
- Fixes #1048
- Fixes #1039
- Fixes #1378
- Fixes #1394
- Fixes #1396

---

#### Internal previews

<details><summary><strong>Toggle expand/collapse</strong></summary><br/>

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/app-host/withdockerfile.md](https://github.com/dotnet/docs-aspire/blob/074e36b1ab2f4f0838e746ba602bfdb56f8dd73f/docs/app-host/withdockerfile.md) | [docs/app-host/withdockerfile](https://review.learn.microsoft.com/en-us/dotnet/aspire/app-host/withdockerfile?branch=pr-en-us-1384) |
| [docs/authentication/keycloak-component.md](https://github.com/dotnet/docs-aspire/blob/074e36b1ab2f4f0838e746ba602bfdb56f8dd73f/docs/authentication/keycloak-component.md) | [docs/authentication/keycloak-component](https://review.learn.microsoft.com/en-us/dotnet/aspire/authentication/keycloak-component?branch=pr-en-us-1384) |
| [docs/caching/stackexchange-redis-caching-overview.md](https://github.com/dotnet/docs-aspire/blob/074e36b1ab2f4f0838e746ba602bfdb56f8dd73f/docs/caching/stackexchange-redis-caching-overview.md) | [docs/caching/stackexchange-redis-caching-overview](https://review.learn.microsoft.com/en-us/dotnet/aspire/caching/stackexchange-redis-caching-overview?branch=pr-en-us-1384) |
| [docs/caching/stackexchange-redis-component.md](https://github.com/dotnet/docs-aspire/blob/074e36b1ab2f4f0838e746ba602bfdb56f8dd73f/docs/caching/stackexchange-redis-component.md) | [docs/caching/stackexchange-redis-component](https://review.learn.microsoft.com/en-us/dotnet/aspire/caching/stackexchange-redis-component?branch=pr-en-us-1384) |
| [docs/caching/stackexchange-redis-distributed-caching-component.md](https://github.com/dotnet/docs-aspire/blob/074e36b1ab2f4f0838e746ba602bfdb56f8dd73f/docs/caching/stackexchange-redis-distributed-caching-component.md) | [docs/caching/stackexchange-redis-distributed-caching-component](https://review.learn.microsoft.com/en-us/dotnet/aspire/caching/stackexchange-redis-distributed-caching-component?branch=pr-en-us-1384) |
| [docs/caching/stackexchange-redis-output-caching-component.md](https://github.com/dotnet/docs-aspire/blob/074e36b1ab2f4f0838e746ba602bfdb56f8dd73f/docs/caching/stackexchange-redis-output-caching-component.md) | [docs/caching/stackexchange-redis-output-caching-component](https://review.learn.microsoft.com/en-us/dotnet/aspire/caching/stackexchange-redis-output-caching-component?branch=pr-en-us-1384) |
| [docs/database/milvus-component.md](https://github.com/dotnet/docs-aspire/blob/074e36b1ab2f4f0838e746ba602bfdb56f8dd73f/docs/database/milvus-component.md) | [docs/database/milvus-component](https://review.learn.microsoft.com/en-us/dotnet/aspire/database/milvus-component?branch=pr-en-us-1384) |
| [docs/fundamentals/components-overview.md](https://github.com/dotnet/docs-aspire/blob/074e36b1ab2f4f0838e746ba602bfdb56f8dd73f/docs/fundamentals/components-overview.md) | [docs/fundamentals/components-overview](https://review.learn.microsoft.com/en-us/dotnet/aspire/fundamentals/components-overview?branch=pr-en-us-1384) |
| [docs/get-started/build-aspire-apps-with-nodejs.md](https://github.com/dotnet/docs-aspire/blob/074e36b1ab2f4f0838e746ba602bfdb56f8dd73f/docs/get-started/build-aspire-apps-with-nodejs.md) | [docs/get-started/build-aspire-apps-with-nodejs](https://review.learn.microsoft.com/en-us/dotnet/aspire/get-started/build-aspire-apps-with-nodejs?branch=pr-en-us-1384) |
| [docs/get-started/build-aspire-apps-with-python.md](https://github.com/dotnet/docs-aspire/blob/074e36b1ab2f4f0838e746ba602bfdb56f8dd73f/docs/get-started/build-aspire-apps-with-python.md) | [docs/get-started/build-aspire-apps-with-python](https://review.learn.microsoft.com/en-us/dotnet/aspire/get-started/build-aspire-apps-with-python?branch=pr-en-us-1384) |
| [docs/messaging/azure-event-hubs-component.md](https://github.com/dotnet/docs-aspire/blob/074e36b1ab2f4f0838e746ba602bfdb56f8dd73f/docs/messaging/azure-event-hubs-component.md) | [docs/messaging/azure-event-hubs-component](https://review.learn.microsoft.com/en-us/dotnet/aspire/messaging/azure-event-hubs-component?branch=pr-en-us-1384) |
| [docs/messaging/azure-web-pubsub-component.md](https://github.com/dotnet/docs-aspire/blob/074e36b1ab2f4f0838e746ba602bfdb56f8dd73f/docs/messaging/azure-web-pubsub-component.md) | [docs/messaging/azure-web-pubsub-component](https://review.learn.microsoft.com/en-us/dotnet/aspire/messaging/azure-web-pubsub-component?branch=pr-en-us-1384) |
| [docs/search/elasticsearch-component.md](https://github.com/dotnet/docs-aspire/blob/074e36b1ab2f4f0838e746ba602bfdb56f8dd73f/docs/search/elasticsearch-component.md) | [docs/search/elasticsearch-component](https://review.learn.microsoft.com/en-us/dotnet/aspire/search/elasticsearch-component?branch=pr-en-us-1384) |

</details>

### CODE REVIEW

File: docs/search/elasticsearch-component.md [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+
+In the _Program.cs_ file of your project, call the `AddElasticsearchClient` extension method to register a `ElasticsearchClient` for use via the dependency injection container. The method takes a connection name parameter.
+
+```csharp
+builder.AddElasticsearchClient("elasticsearch");
+```
+
+## App host usage
+
+To model the RabbitMQ resource in the app host, install the [Aspire.Hosting.Elasticsearch](https://www.nuget.org/packages/Aspire.Hosting.Elasticsearch) NuGet package in the [app host](xref:aspire/app-host) project.
```
Comment (maddymontaquila) [inline docs/search/elasticsearch-component.md:43]: Missed one! Elasticsearch not Rabbit 😂

File: docs/caching/includes/garnet-app-host.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+### [PackageReference](#tab/package-reference)
+
+```xml
+<PackageReference Include="Aspire.Hosting.Garnet"
+                  Version="[SelectVersion]" />
+```
+
+---
+
+In your app host project, register the .NET Aspire Garnet as a resource using the `AddGarnet` method and consume the service using the following methods:
```
Comment (maddymontaquila) [inline docs/caching/includes/garnet-app-host.md:41]: register "the" .NET Aspire Garnet or like.. register the .net aspire garnet extension/component?

File: docs/search/elasticsearch-component.md [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+The .NET Aspire Elasticsearch Client component supports [Microsoft.Extensions.Configuration](/dotnet/api/microsoft.extensions.configuration). It loads the  `ElasticClientsElasticsearchSettings` from configuration by using the `Aspire:Elastic:Clients:Elasticsearch` key. Example _appsettings.json_ that configures some of the options:
+
+```json
+{
+  "Aspire": {
+    "Elastic": {
+      "Clients": {
+        "Elasticsearch": {
+            "ConnectionString": "http://elastic:password@localhost:27011"
```
Comment (Alirexaa) [inline docs/search/elasticsearch-component.md:107]:
```suggestion
            "Endpoint": "http://elastic:password@localhost:27011"
```

File: docs/search/elasticsearch-component.md [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+```csharp
+builder.AddElasticsearchClient(
+    "elasticsearch",
+    settings =>
+        settings.ConnectionString = "http://elastic:password@localhost:27011");
+```
+
+### Use a `CloudId` and an `ApiKey` with configuration providers
+
+When using [Elastic Cloud](https://www.elastic.co/cloud), you can provide the `CloudId` and `ApiKey` in `Aspire:Elastic:Clients:Elasticsearch:Cloud` section and set `Aspire:Elastic:Clients:Elasticsearch:UseCloud` key to `true` when calling `builder.AddElasticsearchClient()`.
```
Comment (Alirexaa) [inline docs/search/elasticsearch-component.md:128]:
```suggestion
When using [Elastic Cloud](https://www.elastic.co/cloud), you can provide the `CloudId` and `ApiKey` in `Aspire:Elastic:Clients:Elasticsearch:Cloud` section when calling `builder.AddElasticsearchClient()`.
```

File: docs/search/elasticsearch-component.md [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        "UseCloud" : true,
+        "Cloud": {
+            "ApiKey": "Valid ApiKey",
+            "CloudId": "Valid CloudId"
+        }
+      }
+    }
+  }
+}
+```
```
Comment (Alirexaa) [inline docs/search/elasticsearch-component.md:150]:
```suggestion
```json
{
  "Aspire": {
    "Elastic": {
      "Clients": {
        "Elasticsearch": {
            "ApiKey": "Valid ApiKey",
            "CloudId": "Valid CloudId"
        }
      }
    }
  }
}
```
```

File: docs/search/elasticsearch-component.md [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+### Use a `CloudId` and an `ApiKey` with inline delegates
+
+```csharp
+builder.AddElasticsearchClient(
+    "elasticsearch",
+    settings => 
+    {
+        settings.UseCloud = true;
+        settings.Cloud.CloudId = "<Valid CloudId>";
+        settings.Cloud.ApiKey = "<Valid ApiKey>";
```
Comment (Alirexaa) [inline docs/search/elasticsearch-component.md:161]:
```suggestion
        settings.CloudId = "<Valid CloudId>";
        settings.ApiKey = "<Valid ApiKey>";
```

**Review by Alirexaa — CHANGES_REQUESTED** [review]
consider looking at https://github.com/dotnet/aspire/issues/4995.

**Comment by mitchdenny** [comment]
@IEvangelist huge thanks for driving this. I'll drive in to help with content tomorrow.

File: docs/database/milvus-component.md [hunk 7]
```diff
@@ -0,0 +1,22 @@
+---
+title: .NET Aspire Milvus database component
+description: This article describes the .NET Aspire Milvus database component.
+ms.topic: how-to
+ms.date: 07/19/2024
+---
+
+# .NET Aspire Milvus database component
```
Comment (mitchdenny) [inline docs/database/milvus-component.md:8]: @timheuer are you able to contiribute content for Milvus?

File: docs/search/elasticsearch-component.md (hunk 3 above)
Comment (IEvangelist) [reply docs/search/elasticsearch-component.md]: Sorry, saw the comment with link to aspire updates before I saw your suggestions. I fixed it.

File: docs/caching/includes/garnet-app-host.md [hunk 8]
```diff
@@ -0,0 +1,56 @@
+<!--
```
Comment (mitchdenny) [inline docs/caching/includes/garnet-app-host.md:1]: Should we remove this commented out stuff and just publish the minimal version?

Comment (mitchdenny) [reply docs/caching/includes/garnet-app-host.md:1]: Oh .. its an include ... never mind.

File: docs/zones/zone-pivot-groups.yml [hunk 9]
```diff
@@ -43,4 +43,13 @@ groups:
       title: Visual Studio Code
     - id: dotnet-cli
       title: .NET CLI
-
+- id: resp-host
+  title: RESP Host
+  prompt: Choose a RESP Hosting resource type
+  pivots:
+    - id: redis
+      title: Stack Exchange Redis
```
Comment (mitchdenny) [inline docs/zones/zone-pivot-groups.yml:51]: This should be Redis, _Stack Exchange Redis_ refers to the client library which would be the same across all of the hosting resource types.

**Review by mitchdenny — APPROVED** [review]

File: docs/zones/zone-pivot-groups.yml (hunk 9 above)
Comment (IEvangelist) [inline docs/zones/zone-pivot-groups.yml:51]:
```suggestion
      title: Redis
```

File: docs/caching/includes/garnet-app-host.md (hunk 8 above)
Comment (IEvangelist) [reply docs/caching/includes/garnet-app-host.md]: I deleted them anyway, they were just comments for us, wouldn't render as HTML.

**Review by jamesmontemagno — APPROVED** [review]
everything you write is magic so didn't even need to read it

**Pull request merged**